File: python/spiral-matrix/spiral_matrix.py
from itertools import cycle


def spiral(size):
    matrix = [[None] * size for _ in range(size)]
    r, c = 0, 0

    deltas = cycle(((0,1), (1,0), (0,-1), (-1,0)))
    dr, dc = next(deltas)

    for i in range(size**2):
        matrix[r][c] = i+1
        if (
            not 0 <= r+dr < size or
            not 0 <= c+dc < size or
            matrix[r+dr][c+dc] is not None
        ):
            dr, dc = next(deltas)
        r += dr
        c += dc
    return matrix

File: lua/spiral-matrix/spiral-matrix.lua
return function(size)
    local matrix = {}
    for x = 1,size do
        matrix[x] = {}
    end

    local x,  y  = 1, 1
    local dx, dy = 0, 1

    for i = 1, size * size do
        matrix[x][y] = i
        if  x + dx < 1 or x + dx > size or
            y + dy < 1 or y + dy > size or
            matrix[x + dx][y + dy] ~= nil
        then
            dx, dy = dy, -dx
        end
        x = x + dx
        y = y + dy
    end

    return matrix
end
